x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/optimizer/LogicalPlanOptimizer.java

```diff
@@ -856,21 +859,35 @@ protected Expression regexToEquals(RegexMatch<?> regexMatch, Literal literal) {
         }
     }
 
-    private static class ReplaceFieldAttributesWithExactSubfield extends OptimizerRules.OptimizerRule<LogicalPlan> {
+    private static class ReplaceFieldAttributesWithExactSubfield extends Rule<LogicalPlan, LogicalPlan> {
 
         @Override
-        protected LogicalPlan rule(LogicalPlan plan) {
-            if (plan instanceof Filter || plan instanceof OrderBy || plan instanceof Aggregate) {
-                return plan.transformExpressionsOnly(FieldAttribute.class, ReplaceFieldAttributesWithExactSubfield::toExact);
-            }
-            return plan;
-        }
+        public LogicalPlan apply(LogicalPlan plan) {
+            Map<EsField, FieldAttribute> esFieldToAttribute = new HashMap<>();
+
+            return plan.transformUp(LogicalPlan.class, p -> {
+                // the source is skipped but used to get the mapping of fields to attributes
+                if (p instanceof EsRelation r) {
+                    for (Attribute out : r.output()) {
+                        // map the field to their actual attributes
+                        if (out instanceof FieldAttribute fa) {
+                            esFieldToAttribute.put(fa.field(), fa);
```

Comment:
question: This code assumes that for each field attribute, there can only be one corresponding `EsField` with that attribute (otherwise, this line would replace the previous field attribute and it would get lost).

Why is that? In the tests that were updated for this PR, I can see that the `job` field has an attribute `job.raw`, which is used in place of `job` after optimization. I wonder if the tests would still work if `job` had additional nested properties `job.prop1`, `job.prop2` etc.

Or, maybe the correspondence `FieldAttribute` -> `EsField` is unique for text fields and we don't care about fields with nested properties etc.? In this case, a small comment here would help understand the assumptions when verifying correctness as a reader.

Reply:
ES fields are just regular fields that as a convention, use a certain name pattern.
Each one can be addressed individually and can have different data types - from the ES perspective is up to the user to figure the relationship between them and use them as such.
Having an attribute to multiple ES fields would create more problems than it would solve.
In terms of documentation - isn't that obvious in how the relationship is designed? FieldAttribute points to one EsField not a list; what else is needed?

Reply:
Thanks, I think I just got confused about the exact meaning of the `FieldAttribute` class.

So, what got me confused is that there's fields, there's properties and there's field attributes. It was not obvious to me from the `FieldAttribute` javadoc that a `FieldAttribute` simply represents an `ESField` as an attribute. It could also be that a `FieldAttribute` represents something belonging to an `ESField`, like any of its properties, and then we'd have a many-to-one relationship which would make the code above faulty. (Also, the constructor of `FieldAttribute` does not construct the attribute _from_ an `ESField`, but just takes it and stores it, so the relationship between field attribs and es fields is not obvious from the `FieldAttribute` class itself.)

If you think that could help avoid more misunderstandings in the future, I could put together a quick doc PR that elaborates the `FieldAttribute` javadoc a bit. What do you think?